src/java/org/apache/cassandra/config/DurationSpec.java

```diff
@@ -24,192 +24,119 @@
+    private DurationSpec(double quantity, TimeUnit unit, TimeUnit minUnit, long max)
     {
-        this(Math.round(quantity), unit);
+        this(Math.round(quantity), unit, minUnit, max);
     }
 
-    public DurationSpec(String value, TimeUnit minUnit)
+    private DurationSpec(String value, TimeUnit minUnit)
     {
-        if (!MAP_UNITS_PER_MIN_UNIT.containsKey(minUnit))
-            throw new ConfigurationException("Invalid smallest unit set for " + value);
-
-        Matcher matcher = TIME_UNITS_PATTERN.matcher(value);
+        Matcher matcher = UNITS_PATTERN.matcher(value);
 
-        if(matcher.find())
+        if (matcher.find())
         {
             quantity = Long.parseLong(matcher.group(1));
             unit = fromSymbol(matcher.group(2));
 
-            if (!MAP_UNITS_PER_MIN_UNIT.get(minUnit).contains(unit))
-                throw new ConfigurationException("Invalid duration: " + value + " Accepted units:" + MAP_UNITS_PER_MIN_UNIT.get(minUnit));
+            //this constructor is used only by extended classes for smallest unit; upper bound is guarded there accordingly
+
+            validateMinUnit(unit, minUnit, value);
         }
         else
         {
-            throw new ConfigurationException("Invalid duration: " + value + " Accepted units:" + MAP_UNITS_PER_MIN_UNIT.get(minUnit) +
+            throw new ConfigurationException("Invalid duration: " + value + " Accepted units:" + acceptedUnits(minUnit) +
                                              " where case matters and only non-negative values.");
         }
     }
 
-    // get vs no-get prefix is not consistent in the code base, but for classes involved with config parsing, it is
-    // imporant to be explicit about get/set as this changes how parsing is done; this class is a data-type, so is
-    // not nested, having get/set can confuse parsing thinking this is a nested type
-    public long quantity()
+    private DurationSpec(String value, TimeUnit minUnit, long max)
     {
-        return quantity;
-    }
+        this (value, minUnit);
 
-    public TimeUnit unit()
-    {
-        return unit;
+        validateMinUnit(unit, minUnit, value);
+        validateQuantity(value, quantity(), unit(), minUnit, max);
     }
 
-    /**
-     * Creates a {@code DurationSpec} of the specified amount of milliseconds.
-     *
-     * @param milliseconds the amount of milliseconds
-     * @return a duration
-     */
-    public static DurationSpec inMilliseconds(long milliseconds)
+    private void validateMinUnit(TimeUnit unit, TimeUnit minUnit, String value)
     {
-        return new DurationSpec(milliseconds, MILLISECONDS);
+        if (unit.compareTo(minUnit) < 0)
+            throw new ConfigurationException(String.format("Invalid duration: %s Accepted units:%s", value, acceptedUnits(minUnit)));
     }
 
-    public static DurationSpec inDoubleMilliseconds(double milliseconds)
+    private String acceptedUnits(TimeUnit minUnit)
```

Comment:
Nit: this method could be `static`
```suggestion
    private static String acceptedUnits(TimeUnit minUnit)
```